### ISSUE TITLE
Multiple wake words possible?

### ISSUE THREAD

**Issue by DieKatzchen** [kahrendt/microWakeWord#19]
Would it be possible to have multiple wake words on the same device?

**Comment by kahrendt** [comment]
Yes, it is possible with two different approaches:
1) The ``micro_wake_word`` component in ESPHome runs two models simultaneously. I have implemented this, but the code needs some clean up. I'm hoping to submit a PR to ESPHome in the next week. You should be able to run two of the current models at the same time, but then the ESP32-S3 is basically at max capacity in terms of CPU usage.
2) Train a single model with multiple wake word probability outputs. I also have this roughly working, but I haven't achieved results that are comparable to single word models (which is to be expected). I want to explore this more in the future with a different model architecture, but it requires some significant changes to the current training process. Currently, all the metrics and related calculations assume a single output indicating the wake word or not.

**Comment by DieKatzchen** [comment]
Approach 1 would be sufficient for my current needs, although as you say it is not ideal.
I assume that with approach 2 it would be possible to know _which_ wakeword was triggered?

**Comment by kahrendt** [comment]
I've made a [draft PR](https://github.com/esphome/esphome/pull/6653) for the ESPHome component that supports running two models at once, so I'm closing this issue.

**Comment by kahrendt** [comment]
Adding multiple wake words into a single model decreases accuracy significantly, at least with the same number of parameters. The current MixedNet. Different architectures may be more efficient at it, but it isn't something I've explored heavily. The MixedNet architecture is quite similar to ``ds_tc_resnet`` in a lot of ways, where it mainly uses depthwise convolutions. If you only provide 1 kernel size per layer with the ``mixconv_kernel_sizes`` parameter, then it will be exactly a single depthwise convolution.

Training multiple wake words in a single model requires quite a bit of refactoring since each validation batch assumes there is a binary output when computing the estimated false accepts per hour.

**Comment by rolyantrauts** [comment]
A wakeword model is essentially an image detection model and is no different from Yolo but with audio streams having and requiring a much faster sample rate wakeword are optimised for size using techniques such as MFCC and streaming interfaces.
If you take the COCO dataset to train many image models its 80 classes are essentially 'wakeword' and that is all you do is add another class that will provide a very slight increase in parameters, but will not effect accuracy significantly unless the wakeword have a large cross-entropy by being very similar.
The keyword spotting standard benchmark, Google speech command datasets v1 and v2 contain multiple classes and is default dataset for testing multiple keyword accuracy that similar sounding syllable such as go & no can reduce accuracy but selection of unique syllable wakeword will have little to no accuracy effect, in fact the opposite as many false positives are due to a near binary dataset of class imbalance that can be seen during a hockey stick training curve result due to lack of cross entropy of too few classes.
